### ISSUE TITLE
Ingest date processor: Support nanosecond precision with UNIX format

### ISSUE THREAD

**Issue by andrewkroh** [elastic/elasticsearch#86106]
### Description

I would like to be able to parse unix epoch timestamps that contain a fractional value with nanosecond precision (e.g. `1647479604.334549372`). These currently parse, but only millisecond precision is retained. Given that the date processor supports nanosecond output formats, I think it would be good to retain the nanosecond precision.

```
POST _ingest/pipeline/_simulate
{
  "pipeline": {
    "processors": [
      {
        "date": {
          "field": "message",
          "formats": [
            "UNIX"
          ],
          "output_format": "yyyy-MM-dd'T'HH:mm:ss.SSSSSSSSSXXX"
        }
      }
    ]
  },
  "docs": [
    {
      "_source": {
        "message": "1647479604.334549372"
      }
    }
  ]
}
```

Output:

```
{
  "docs" : [
    {
      "doc" : {
        "_index" : "_index",
        "_id" : "_id",
        "_source" : {
          "@timestamp" : "2022-03-17T01:13:24.334000000Z", # NOTE: This only contains millisecond precision.
          "message" : "1647479604.334549372"
        },
        "_ingest" : {
          "timestamp" : "2022-04-23T02:15:28.600856847Z"
        }
      }
    }
  ]
}
```

My assumption is that the cast to long truncates the everything after the millis at

https://github.com/elastic/elasticsearch/blob/ce6d2e7f31ab7c7c36cc49eb708549b64f68086a/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java#L49

**Comment by HiDAl** [comment]
Hi @andrewkroh 

First, sorry for the late response. 
Second, you're right, sadly, the [`Date processor`](https://www.elastic.co/guide/en/elasticsearch/reference/current/date-processor.html) does not support input date with fractional nanoseconds. So, it's impossible to transform using the built-in Date processor... but, we have a workaround that could be less efficient, but it does the work. Basically, you can use a `Script processor` that using custom logic parses your date. I've written the following basic example:

```
POST _ingest/pipeline/_simulate
{
  "pipeline": {
    "processors": [
      {
        "script": {
          "lang": "painless",
          "source": "def parts = ctx['message'].splitOnToken('.'); ctx.parsedEpoch = Instant.ofEpochSecond(Long.parseLong(parts[0]), Long.parseLong(parts[1])).atZone(ZoneOffset.UTC);"
        }
      },
      {
        "date": {
          "field": "parsedEpoch",
          "formats": [
            "ISO8601"
          ],
          "output_format": "strict_date_optional_time_nanos"
        }
      }
    ]
  },
  "docs": [
    {
      "_source": {
        "message": "1647479604.334549372"

      }
    }
  ]
}

{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_version": "-3",
        "_source": {
          "message": "1647479604.334549372",
          "parsedEpoch": "2022-03-17T01:13:24.334549372Z",
          "@timestamp": "2022-03-17T01:13:24.334549372Z"
        },
        "_ingest": {
          "timestamp": "2023-06-21T07:56:45.73386Z"
        }
      }
    }
  ]
}
```

Basically:

1. takes the field `message`
2. split it into parts
3. using Java [`Instant.ofEpochSecond`](https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html#ofEpochSecond-long-long-), it creates a new field (`parsedEpoch`)
4. Later, the same field is parsed using another `Date processor` just to show that the output is valid (step not needed at all).

I'll keep this ticket open, because, even though we have a workaround, we have to decide as a team whether makes sense to implement support for this kind of dates

**Comment by HiDAl** [comment]
Hello @andrewkroh,

we have discussed this issue internally and even though it'd be nice to have support for nanoseconds, it's not at the top of our priorities right now. That's why we have marked it as `good first issue` in case someone new to the company or external contributors wants to implement it.

**Comment by joegallo** [comment]
`epoch_second` supports the desired behavior (see #99388 for a little more context on why I don't want to add a `UNIX_NS`, for example).
